File: 59-SpiralMatrixII.cpp
/*

Given an integer n, generate a square matrix filled with elements from 1 to n2 in spiral order.

For example,
Given n = 3,

You should return the following matrix:
[
 [ 1, 2, 3 ],
 [ 8, 9, 4 ],
 [ 7, 6, 5 ]
]

*/


class Solution {

public:

    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n,vector<int>(n));

        for(int num = 0,left = 0, right = n-1, top = 0, bottom = n-1;
            left <= right && top <= bottom;
            ++left, --right, ++top, --bottom) {

            for (int j = left; j <= right; ++j) {
                res[top][j] = ++num;
            }
            for (int i = top+1; i <= bottom; ++i) {
                res[i][right] = ++num;
            }
            for (int j = right-1; top < bottom && j >= left; --j) {
                res[bottom][j] = ++num;
            }
            for (int i = bottom-1; left < right && i > top; --i) {
                res[i][left] = ++num;
            }
        }
        return res;
    }
};

File: java/59-SpiralMatrixII.java
/*

Given an integer n, generate a square matrix filled with elements from 1 to n2 in spiral order.

For example,
Given n = 3,

You should return the following matrix:
[
 [ 1, 2, 3 ],
 [ 8, 9, 4 ],
 [ 7, 6, 5 ]
]

*/


class Solution {
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];

        int l = 0, r = n-1;
        int t = 0, b = n-1;
        int num = 0;
        
        while(l<=r && t<=b) {
            for(int j = l; j<=r; j++) matrix[t][j] = ++num;
            for(int i = t+1; i<=b; i++) matrix[i][r] = ++num;
            for(int j = r-1; t!=b&&j>=l; j--) matrix[b][j] = ++num;
            for(int i = b-1; l!=r && i>t; i--) matrix[i][l] = ++num;
            ++t;
            --b;
            ++l;
            --r;
        }
        return matrix;
        
    }
}
